core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java

```diff
@@ -232,4 +239,22 @@ private void putIfNotNull(Map<String, String> context, String variable, String k
         }
     }
 
+    public String changePrimaryGateway(Stack stack) throws CloudbreakException {
+        GatewayConfig formerPrimaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+        List<GatewayConfig> gatewayConfigs = gatewayConfigService.getAllGatewayConfigs(stack);
+        Optional<GatewayConfig> newPrimaryCandidate = gatewayConfigs.stream().filter(gc -> !gc.isPrimary()).findFirst();
+        if (newPrimaryCandidate.isPresent()) {
+            GatewayConfig newPrimary = newPrimaryCandidate.get();
+            Set<Node> allNodes = collectNodes(stack);
+            try {
+                hostOrchestratorResolver.get(stack.getOrchestrator().getType()).changePrimaryGateway(formerPrimaryGatewayConfig, newPrimary, gatewayConfigs,
+                        allNodes, clusterDeletionBasedExitCriteriaModel(stack.getId(), stack.getCluster().getId()));
+                return newPrimary.getHostname();
+            } catch (CloudbreakOrchestratorException ex) {
+                throw new CloudbreakException(ex);
+            }
+        } else {
+            throw new CloudbreakException("Primary gateway change is not possible!");
```

Comment:
Shouldn't we have to say 'Primary gateway change is not possible because there is no available node for the action'.